src/main/cpp/util/profiler.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BAZEL_SRC_MAIN_CPP_UTIL_PROFILER_H_
+#define BAZEL_SRC_MAIN_CPP_UTIL_PROFILER_H_
+
+#include <stdint.h>  // int64_t
+#include <stdio.h>   // printf
+
+namespace blaze_util {
+namespace profiler {
+
+// A time duration, measured in some implementation-dependent units.
+//
+// Using a struct to wrap int64_t yields a unique type that we can't
+// accidentally use as a type-synonym of other int64_t types.
+typedef struct _Ticks {
+  int64_t value_;
+  static _Ticks Now();
+} Ticks;
+
+// A time duration, measured in milliseconds.
+//
+// Using a struct to wrap int64_t yields a unique type that we can't
+// accidentally use as a type-synonym of other int64_t types.
+typedef struct _Duration {
+  int64_t millis_;
```

Comment:
Why not count microseconds? I can imagine that some operations that you want to measure are done very frequently and take less than a millisecond each. I suppose `clock_gettime()` is precise enough for that to work; what does `clock_getres()` return on Linux?

Reply:
On my Debian Rodete `clock_getres()` reports 1ns resolution.
On my Windows the resolution is about 290ns (`QueryPerformanceFrequency` reports 3409750 / sec).

Reply:
I didn't intend to carry out microbenchmarks, I'm more interested in low-hanging-fruit optimizations. For that, millisecond resolution is fine.